arraycontext/pytest.py

```diff
@@ -224,6 +224,26 @@ def __str__(self):
         return "<PytatoJAXArrayContext>"
 
 
+class _PytestCupyArrayContextFactory(PytestArrayContextFactory):
+    @classmethod
+    def is_available(cls) -> bool:
+        try:
+            import cupy  # type: ignore[import-untyped]  # noqa: F401
+            return True
+        except ImportError:
+            return False
+
+    def __call__(self):
+        from arraycontext import CupyArrayContext
+        return CupyArrayContext()
+
+    def __str__(self):
+        import cupy  # pylint: disable=import-error
+        d = cupy.cuda.runtime.getDeviceProperties(cupy.cuda.Device())
+        name = d["name"].decode("utf-8")
+        return f"<CupyArrayContext> on {cupy.cuda.Device()}:{name}"
```

Comment:
This is the only part of the code that give me pause for supporting something like [cupy-rocm](https://docs.cupy.dev/en/latest/install.html#using-cupy-on-amd-gpu-experimental).